3.next - Add RouteBuilder::loadPlugin.

markstory opened the pull request:
Deprecate the static method and provide a route builder method that allows routes to be loaded inside scopes.

Refs #10689

markstory committed:
Add RouteBuilder::loadPlugin.

Deprecate the static method and provide a route builder method that
allows routes to be loaded inside scopes.

Refs #10689

ravage84 commented on src/Routing/RouteBuilder.php:
```diff
@@ -519,6 +521,35 @@ protected function _methodRoute($method, $template, $target, $name)
     /**
+     * Load routes from a plugin.
+     *
+     * The routes file will have a local variable named `$routes` made available which contains
```
will -> must?

markstory replied:
I was trying to say the routes file will have access to `$routes` as a local variable.

ravage84 replied:
Yeah, I see now. Never mind.

ravage84 commented on src/Routing/RouteBuilder.php:
```diff
@@ -519,6 +521,35 @@ protected function _methodRoute($method, $template, $target, $name)
+     * the current RouteBuilder instance.
+     *
+     * @param string $name The plugin name
+     * @param string $file The routes file to load. Defaults to `routes.php`
```
👍

ravage84 commented on src/Routing/RouteBuilder.php:
```diff
@@ -519,6 +521,35 @@ protected function _methodRoute($method, $template, $target, $name)
+        if (!Plugin::loaded($name)) {
+            throw new MissingPluginException(['plugin' => $name]);
+        }
+        $path = Plugin::configPath($name) . DIRECTORY_SEPARATOR . $file;
```
I'd insert a blank line before for readability.

ravage84 commented on src/Routing/RouteBuilder.php:
```diff
@@ -519,6 +521,35 @@ protected function _methodRoute($method, $template, $target, $name)
+                $path
+            ));
+        }
+        $routes = $this;
```
I'd insert a blank line before for readability.

ravage84 commented:
@markstory can you elaborate what the implications are for the future app setup?

markstory commented:
@ravage84 I would like to see `Plugin::routes()` go away, and instead plugins either use events to automagically add their routes, or the installation of plugins requires adding `$routes->loadPlugin(...)` in the right place.

ravage84 commented:
@markstory I see. I just "fear" the installation of plugins or better said the configuration of plugin routes could become more complicated because of this...

markstory committed:
Blank lines.

markstory commented:
@ravage84 Fully removing `Plugin::routes()` and only having this method would make plugin installation more complex, which is not my goal. I want to make routing easier to understand and simpler to change. I found that `Plugin::routes()` is quite opaque and makes routing harder to understand and change.

Perhaps `Plugin::routes()` can live alongside this new method as a simple way to splat all the routes at once, while this new method gives people finer grained control over routes.

markstory committed:
Undeprecate Plugin::routes()

We'll leave it alone for now as it makes an easy to use API for simple
plugin setups.

lorenzo commented:
@markstory what happens if you use both `loadPlugin` and `Plugin::routes()`. Should this be an error condition?

markstory commented:
@lorenzo The routes would be loaded twice. Which could result in incorrect URLs being generated. While I like to use `include_once` for routes to prevent that, but that complicates testing as we need to reload routes quite often.

If the routes are named, they will hit an error, but un-named routes are harder to catch.
